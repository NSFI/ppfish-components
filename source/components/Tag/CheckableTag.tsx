import * as React from "react";
import classNames from "classnames";

export interface CheckableTagProps {
  prefixCls?: string;
  className?: string;
  checked: boolean;
  onChange?: (checked: boolean) => void;
}

const CheckableTag:React.FC<CheckableTagProps> = props => {

  const { prefixCls = "fishd-tag", className, checked, onChange, ...restProps } = props;
  const cls = classNames(
    prefixCls,
    {
      [`${prefixCls}-checkable`]: true,
      [`${prefixCls}-checkable-checked`]: checked
    },
    className
  );

  const handleClick = () => {
    const { checked, onChange } = props;
    if (onChange) {
      onChange(!checked);
    }
  };
  return <div {...restProps} className={cls} onClick={handleClick} />;
};
export default CheckableTag;
