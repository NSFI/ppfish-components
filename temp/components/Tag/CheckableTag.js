var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import classNames from 'classnames';
var CheckableTag = /** @class */ (function (_super) {
    __extends(CheckableTag, _super);
    function CheckableTag() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleClick = function () {
            var _a = _this.props, checked = _a.checked, onChange = _a.onChange;
            if (onChange) {
                onChange(!checked);
            }
        };
        return _this;
    }
    CheckableTag.prototype.render = function () {
        var _a;
        var _b = this.props, _c = _b.prefixCls, prefixCls = _c === void 0 ? 'fishd-tag' : _c, className = _b.className, checked = _b.checked, restProps = __rest(_b, ["prefixCls", "className", "checked"]);
        var cls = classNames(prefixCls, (_a = {},
            _a[prefixCls + "-checkable"] = true,
            _a[prefixCls + "-checkable-checked"] = checked,
            _a), className);
        delete restProps.onChange; // TypeScript cannot check delete now.
        return React.createElement("div", __assign({}, restProps, { className: cls, onClick: this.handleClick }));
    };
    return CheckableTag;
}(React.Component));
export default CheckableTag;
