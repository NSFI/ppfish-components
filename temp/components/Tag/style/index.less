@import (reference) "../../../assets/css/index";

@tag-prefix-cls: ~"@{fishd-prefix}-tag";

/* stylelint-disable declaration-no-important */
.@{tag-prefix-cls} {
  .reset-component;
  display: inline-flex;
  align-items: center;
  // line-height: 20px;
  height: 20px;
  padding: 0 8px;
  border-radius: @border-radius-base;
  // border: @border-width-base @border-style-base @border-color-base;
  background: @tag-default-bg;
  font-size: @tag-font-size;
  transition: all 0.3s @ease-out;
  opacity: 1;
  margin-right: 8px;
  cursor: pointer;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  &:hover {
    opacity: 0.85;
    .@{iconfont-prefix}-close-modal-line.invisible {
      opacity: 1;
    }
  }
  &:only-child {
    margin-right: 0;
  }

  &,
  a,
  a:hover {
    color: @tag-default-color;
  }

  > a:first-child:last-child {
    display: inline-block;
    // margin: 0 -8px;
    // padding: 0 8px;
  }

  .@{iconfont-prefix}-close-modal-line.invisible {
    opacity: 0;
  }

  .@{iconfont-prefix}-close-modal-line {
    .iconfont-size-under-12px(8px);
    cursor: pointer;
    margin-left: 3px;
    transition: all 0.3s;
    color: @text-color-secondary;
    font-weight: bold;

    &:hover {
      color: @heading-color;
    }
  }

  &-has-color {
    border-color: transparent;
    &,
    a,
    a:hover,
    .@{iconfont-prefix}-close-modal-line,
    .@{iconfont-prefix}-close-modal-line:hover {
      color: #fff;
    }
  }

  &-checkable {
    background-color: transparent;
    border-color: transparent;

    &:not(&-checked):hover {
      color: @primary-color;
    }

    &:active,
    &-checked {
      color: #fff;
    }

    &-checked {
      background-color: @primary-6;
    }

    &:active {
      background-color: @primary-7;
    }
  }

  &-close {
    width: 0 !important;
    padding: 0;
    margin: 0;
  }

  &-zoom-enter,
  &-zoom-appear {
    animation: fishdFadeIn 0.2s @ease-in-out-circ;
    animation-fill-mode: both;
  }

  &-zoom-leave {
    animation: fishdZoomOut 0.3s @ease-in-out-circ;
    animation-fill-mode: both;
  }

  // @colors: pink, magenta, red, volcano, orange, yellow, gold, cyan, lime, green, blue, geekblue, purple;

  // // mixin to iterate over colors and create CSS class for each one
  // .make-color-classes(@i: length(@colors)) when (@i > 0) {
  //   .make-color-classes(@i - 1);
  //   @color: extract(@colors, @i);
  //   @lightColor: "@{color}-1";
  //   @lightBorderColor: "@{color}-3";
  //   @darkColor: "@{color}-6";
  //   &-@{color} {
  //     color: @@darkColor;
  //     background: @@lightColor;
  //     border-color: @@lightBorderColor;
  //   }
  //   &-@{color}-inverse {
  //     background: @@darkColor;
  //     border-color: @@darkColor;
  //     color: #fff;
  //   }
  // }

  // .make-color-classes();
}
/* stylelint-enable */
