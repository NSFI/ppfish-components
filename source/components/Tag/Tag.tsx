import * as React from "react";
import classNames from "classnames";
import omit from "omit.js";
import Icon from "../Icon";
import Animate from "rc-animate";
import CheckableTag from "./CheckableTag";
import ReactDOM from "react-dom";

export interface TagProps extends React.HTMLAttributes<HTMLDivElement> {
  prefixCls?: string;
  className?: string;
  color?: string;
  /** 标签是否可以关闭 */
  closable?: boolean;
  visible?: boolean;
  autoShowClose?: boolean;
  /** 关闭时的回调 */
  onClose?: Function;
  /** 动画关闭后的回调 */
  afterClose?: Function;
  style?: React.CSSProperties;
}

export interface TagState {
  closing: boolean;
  closed: boolean;
  visible: boolean;
}

const BaseTag: React.FC<TagProps> = props => {
  const {
    prefixCls,
    closable,
    color,
    className,
    children,
    style,
    autoShowClose,
    ...otherProps
  } = props;

  const [stateClosing, setStateClosing] = React.useState<TagState["closing"]>(false);
  const [stateClosed, setStateClosed] = React.useState<TagState["closed"]>(false);
  const [stateVisible, setStateVisible] = React.useState<TagState["visible"]>(true);

  const tagRef = React.createRef<HTMLElement>();

  React.useEffect(() => {
    if ("visible" in otherProps) {
      setStateVisible(otherProps.visible);
    }
  }, [otherProps.visible]);

  React.useEffect(() => {
    if (!stateVisible) {
      close();
    } else {
      show();
    }
  }, [stateVisible]);

  const handleIconClick = (e: React.MouseEvent<HTMLElement>) => {
    const {onClose} = props;
    if (onClose) {
      onClose(e);
    }
    if (e.defaultPrevented || "visible" in props) {
      return;
    }
    setStateVisible(false);
  };

  const close = () => {
    if (stateClosing || stateClosed) {
      return;
    }
    const dom = ReactDOM.findDOMNode(tagRef.current) as HTMLElement;
    dom.style.width = `${dom.getBoundingClientRect().width}px`;
    setStateClosing(true);
  };

  const show = () => {
    setStateClosed(false);
  };

  const animationEnd = (_: string, existed: boolean) => {
    if (!existed && !stateClosed) {

      setStateClosed(true);
      setStateClosing(false);

      const afterClose = props.afterClose;
      if (afterClose) {
        afterClose();
      }
    } else {
      setStateClosed(false);
    }
  };

  const classString = classNames(
    prefixCls,
    {
      [`${prefixCls}-has-color`]: color,
      [`${prefixCls}-close`]: stateClosing
    },
    className
  );
  // fix https://fb.me/react-unknown-prop
  const divProps = omit(otherProps, ["onClose", "afterClose", "visible"]);
  const tagStyle = {
    backgroundColor: color ? color : null,
    ...style
  };
  let closeIcon = null;
  if (closable) {
    if (autoShowClose) {
      closeIcon = <Icon type="close-modal-line" onClick={handleIconClick} />;
    } else {
      closeIcon = (
        <Icon type="close-modal-line" onClick={handleIconClick} className="invisible" />
      );
    }
  }

  const tag = stateClosed ? null : (
    <div data-show={!stateClosing} {...divProps} className={classString} style={tagStyle} ref={tagRef}>
      {children}
      {closeIcon}
    </div>
  );

  return (
    <Animate
      component=""
      showProp="data-show"
      transitionName={`${prefixCls}-zoom`}
      transitionAppear
      onEnd={animationEnd}
    >
      {tag}
    </Animate>
  );
};

type BaseTagType = typeof BaseTag;

interface TagInterface extends BaseTagType {
  CheckableTag: typeof CheckableTag;
}

const Tag = BaseTag as TagInterface;

Tag.displayName = 'Tag';

Tag.defaultProps = {
  prefixCls: "fishd-tag",
  closable: false,
  autoShowClose: true,
};

Tag.CheckableTag = CheckableTag;

export default Tag;
