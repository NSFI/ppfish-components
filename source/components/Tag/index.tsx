import Tag from "./Tag";
import CheckableTag from "./CheckableTag";
import './style/index.less';

export { CheckableTagProps } from './CheckableTag';
Tag.CheckableTag = CheckableTag;

export default Tag;
