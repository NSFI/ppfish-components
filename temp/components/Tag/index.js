var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import Animate from 'rc-animate';
import classNames from 'classnames';
import omit from 'omit.js';
import { polyfill } from 'react-lifecycles-compat';
import Icon from '../Icon/index';
import CheckableTag from './CheckableTag';
import './style/index.less';
var Tag = /** @class */ (function (_super) {
    __extends(Tag, _super);
    function Tag() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            closing: false,
            closed: false,
            visible: true
        };
        _this.handleIconClick = function (e) {
            var onClose = _this.props.onClose;
            if (onClose) {
                onClose(e);
            }
            if (e.defaultPrevented || 'visible' in _this.props) {
                return;
            }
            _this.setState({ visible: false });
        };
        _this.close = function () {
            if (_this.state.closing || _this.state.closed) {
                return;
            }
            var dom = ReactDOM.findDOMNode(_this);
            dom.style.width = dom.getBoundingClientRect().width + "px";
            // It's Magic Code, don't know why
            dom.style.width = dom.getBoundingClientRect().width + "px";
            _this.setState({
                closing: true
            });
        };
        _this.show = function () {
            _this.setState({
                closed: false
            });
        };
        _this.animationEnd = function (_, existed) {
            if (!existed && !_this.state.closed) {
                _this.setState({
                    closed: true,
                    closing: false
                });
                var afterClose = _this.props.afterClose;
                if (afterClose) {
                    afterClose();
                }
            }
            else {
                _this.setState({
                    closed: false
                });
            }
        };
        return _this;
    }
    Tag.getDerivedStateFromProps = function (nextProps) {
        return 'visible' in nextProps ? { visible: nextProps.visible } : null;
    };
    Tag.prototype.componentDidUpdate = function (_prevProps, prevState) {
        if (prevState.visible && !this.state.visible) {
            this.close();
        }
        else if (!prevState.visible && this.state.visible) {
            this.show();
        }
    };
    Tag.prototype.isPresetColor = function (color) {
        return false;
        // if (!color) { return false; }
        // return (
        //   /^(pink|red|yellow|orange|cyan|green|blue|purple|geekblue|magenta|volcano|gold|lime)(-inverse)?$/
        //   .test(color)
        // );
    };
    Tag.prototype.render = function () {
        var _a;
        var _b = this.props, prefixCls = _b.prefixCls, closable = _b.closable, color = _b.color, className = _b.className, children = _b.children, style = _b.style, autoShowClose = _b.autoShowClose, otherProps = __rest(_b, ["prefixCls", "closable", "color", "className", "children", "style", "autoShowClose"]);
        var isPresetColor = this.isPresetColor(color);
        var classString = classNames(prefixCls, (_a = {},
            _a[prefixCls + "-" + color] = isPresetColor,
            _a[prefixCls + "-has-color"] = color && !isPresetColor,
            _a[prefixCls + "-close"] = this.state.closing,
            _a), className);
        // fix https://fb.me/react-unknown-prop
        var divProps = omit(otherProps, ['onClose', 'afterClose', 'visible']);
        var tagStyle = __assign({ backgroundColor: color && !isPresetColor ? color : null }, style);
        var closeIcon = null;
        if (closable) {
            if (autoShowClose) {
                closeIcon = React.createElement(Icon, { type: "close-modal-line", onClick: this.handleIconClick });
            }
            else {
                closeIcon = (React.createElement(Icon, { type: "close-modal-line", onClick: this.handleIconClick, className: "invisible" }));
            }
        }
        var tag = this.state.closed ? null : (React.createElement("div", __assign({ "data-show": !this.state.closing }, divProps, { className: classString, style: tagStyle }),
            children,
            closeIcon));
        return (React.createElement(Animate, { component: "", showProp: "data-show", transitionName: prefixCls + "-zoom", transitionAppear: true, onEnd: this.animationEnd }, tag));
    };
    Tag.CheckableTag = CheckableTag;
    Tag.defaultProps = {
        prefixCls: 'fishd-tag',
        closable: false,
        autoShowClose: true
    };
    return Tag;
}(React.Component));
polyfill(Tag);
export default Tag;
