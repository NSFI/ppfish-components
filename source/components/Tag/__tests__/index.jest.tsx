import React from 'react';
import { mount } from 'enzyme';
import Tag from '../index';

describe('Tag', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should be closable', () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <Tag closable onClose={onClose} />
    );
    expect(wrapper.find('.fishdicon-close-modal-line').length).toBe(1);
    expect(wrapper.find('.fishd-tag').length).toBe(1);
    wrapper.find('.fishdicon-close-modal-line').simulate('click');
    expect(onClose).toBeCalled();
    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.fishd-tag').length).toBe(0);
  });

  it('should not be closed when prevent default', () => {
    const onClose = (e) => {
      e.preventDefault();
    };
    const wrapper = mount(
      <Tag closable onClose={onClose} />
    );
    expect(wrapper.find('.fishdicon-close-modal-line').length).toBe(1);
    expect(wrapper.find('.fishd-tag').length).toBe(1);
    wrapper.find('.fishdicon-close-modal-line').simulate('click');
    jest.runAllTimers();
    expect(wrapper.find('.fishd-tag').length).toBe(1);
  });

  it('can be controlled by visible', () => {
    const wrapper = mount(
      <Tag visible />
    );
    expect(wrapper.render()).toMatchSnapshot();
    wrapper.setProps({ visible: false });
    jest.runAllTimers();
    expect(wrapper.render()).toMatchSnapshot();
    wrapper.setProps({ visible: true });
    jest.runAllTimers();
    expect(wrapper.render()).toMatchSnapshot();
  });
});
