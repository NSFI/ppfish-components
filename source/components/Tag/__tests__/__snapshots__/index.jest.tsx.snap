// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag can be controlled by visible 1`] = `
<div
  class="fishd-tag fishd-tag-zoom-appear"
  data-show="true"
/>
`;

exports[`Tag can be controlled by visible 2`] = `
<div
  class="fishd-tag fishd-tag-close"
  data-show="false"
  style="width: 0px;"
/>
`;

exports[`Tag can be controlled by visible 3`] = `null`;
